nixos/modules/system/boot/luksroot.nix

```
@@ -206,19 +206,18 @@ let
 
     # LUKS
     open_normally() {
-        ${if (dev.keyFile != null) then ''
+        ${optionalString (dev.keyFile != null) ''
         if wait_target "key file" ${dev.keyFile}; then
             ${csopen} --key-file=${dev.keyFile} \
               ${optionalString (dev.keyFileSize != null) "--keyfile-size=${toString dev.keyFileSize}"} \
-              ${optionalString (dev.keyFileOffset != null) "--keyfile-offset=${toString dev.keyFileOffset}"}
+              ${optionalString (dev.keyFileOffset != null) "--keyfile-offset=${toString dev.keyFileOffset}"} \
+                && return 0
         else
             ${if dev.fallbackToPassword then "echo" else "die"} "${dev.keyFile} is unavailable"
-            echo " - failing back to interactive password prompt"
-            do_open_passphrase
         fi
-        '' else ''
-        do_open_passphrase
+        echo " - failing back to interactive password prompt"
         ''}
+        do_open_passphrase
```

```suggestion
        do_open_passphrase
        ''}
```
Shouldn't this be like this?

---

No, because that closes the conditional section for `dev.keyFile != null`. The `open_normally` should always contain at least that ask for password.

The open using key either succeeds and thus `return 0` or it fallbacks to `do_open_passphase`. When there is no key it either dies or echoes a message and again goes for `do_open_passphase`. Either way it asks for password as a fallback. That is why I took the `do_open_passphase` outside of the `${}`.

The primary change here is that there was a path when `keyFile` was set and there was a file (but possibly invalid) the open would just simply fail. With this every fail (except when `fallbackToPassword` is set) results in `do_open_passphase`  and thus is no longer in condition.

The suggested change would result in ask for passhpase happening only when `keyFile` is set but we should ask for it especially when `keyFile` is not set and of course as a fallback when open using key file fails.

---

@SuperSandro2000 are you ok with this explanation?

---

I am not familiar with the code, I just did a best guess.